Add basic command line support (#40)

nocive opened the pull request:
Add command line support to rollbar notifier, addresses #40 

- [x] collect ```argv``` and send it in ```build_server_data()``` (must have php directive ```register_argc_argv``` on, otherwise it'll not be set)
- [x] send extra ```php_context``` variable for all errors, indicating the context in which the error happened (```http``` or ```cli```)
- [x] add test coverage

nocive commented:
cc @Crisfole

Crisfole commented:
@nocive: I see this, the changes look really small and don't break any tests.  Any chance you have a demo of how this would work? (That demonstrates it works, for instance?)

nocive commented:
sure, will work on that, was just aiming to get some early feedback if this would be the way to go about implemeting it ;)

Crisfole commented:
@nocive This looks good to me I just want to see an example and I think I'll ear mark it for a merge...We've got lots of merging coming up in the next few days I think...we're long overdue.

nocive commented:
@Crisfole here's a payload sample using the async/relayed rollbar setup with this exact same code running:

```
{"data":{"timestamp":1460564335,"environment":"test","level":"error","language":"php","framework":"php","notifier":{"name":"rollbar-php","version":"0.15.0"},"uuid":"973bc414-b357-4b19-b4d1-dafb084dbdb7","body":{"trace":{"frames":[{"filename":"##omitted##","lineno":31,"method":"<main>"},{"filename":"##omitted##","lineno":126,"method":"run"},{"filename":"##omitted##","lineno":96,"method":"doRun"},{"filename":"##omitted##","lineno":195,"method":"doRun"},{"filename":"##omitted##","lineno":882,"method":"doRunCommand"},{"filename":"##omitted##","lineno":257,"method":"run"},{"filename":"##omitted##","lineno":24,"method":"execute"}],"exception":{"class":"Exception","message":"rollbar error simulation"}}},"server":{"host":"##omitted##","argv":["somescript","-vvvv","test:command","--arg=value"]},"person":null},"access_token":"##omitted##"}
```

and pretty printed:
```
{
  "data": {
    "timestamp": 1460564335,
    "environment": "test",
    "level": "error",
    "language": "php",
    "framework": "php",
    "notifier": {
      "name": "rollbar-php",
      "version": "0.15.0"
    },
    "uuid": "973bc414-b357-4b19-b4d1-dafb084dbdb7",
    "body": {
      "trace": {
        "frames": [
          {
            "filename": "##omitted##",
            "lineno": 31,
            "method": "<main>"
          },
          {
            "filename": "##omitted##",
            "lineno": 126,
            "method": "run"
          },
          {
            "filename": "##omitted##",
            "lineno": 96,
            "method": "doRun"
          },
          {
            "filename": "##omitted##",
            "lineno": 195,
            "method": "doRun"
          },
          {
            "filename": "##omitted##",
            "lineno": 882,
            "method": "doRunCommand"
          },
          {
            "filename": "##omitted##",
            "lineno": 257,
            "method": "run"
          },
          {
            "filename": "##omitted##",
            "lineno": 24,
            "method": "execute"
          }
        ],
        "exception": {
          "class": "Exception",
          "message": "rollbar error simulation"
        }
      }
    },
    "server": {
      "host": "##omitted##",
      "argv": [
        "somescript",
        "-vvvv",
        "test:command",
        "--arg=value"
      ]
    },
    "person": null
  },
  "access_token": "##omitted##"
}

nocive commented:
I'm also thinking about including the ```context``` variable as a server argument, so that you can clearly see from the rollbar backend and notifications what the error execution context is (possible values: http or cli).

brianr commented on src/rollbar.php:
```diff
@@ -122,6 +122,7 @@ class RollbarNotifier {
         'scrub_fields', 'shift_function', 'timeout', 'report_suppressed', 'use_error_reporting', 'proxy');
 
     // cached values for request/server/person data
+    private $_context = null;
```
Sorta minor, but can you call this variable something besides `_context`? One of the fields that the Rollbar API understands is also "context" and I worry that this could be confused with that.

Just as some ideas, maybe `_execution_context` or `_php_context` ?

nocive replied:
Sure, was actually planning to do it alrdy, check the pr description :)
I'll be happy to make any changes you guys have to proppose or implement
other ideas.

On Wed, Apr 13, 2016, 23:04 Brian Rue <notifications@github.com> wrote:

> In src/rollbar.php
> <https://github.com/rollbar/rollbar-php/pull/80#discussion_r59625282>:
>
> > @@ -122,6 +122,7 @@ class RollbarNotifier {
> >          'scrub_fields', 'shift_function', 'timeout', 'report_suppressed', 'use_error_reporting', 'proxy');
> >
> >      // cached values for request/server/person data
> > +    private $_context = null;
>
> Sorta minor, but can you call this variable something besides _context?
> One of the fields that the Rollbar API understands is also "context" and I
> worry that this could be confused with that.
>
> Just as some ideas, maybe _execution_context or _php_context ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rollbar/rollbar-php/pull/80/files/deb80620fecb2fa78ce2d0c260a5ace744f8f6d5#r59625282>
>

nocive commented:
@brianr done, also added ```php_context``` to ```server_data``` and fixed a couple of indentation issues.

brianr commented:
@nocive awesome - looks good to me.

Last thing: is it feasible to get some test coverage for this?

nocive commented:
Sure, also in my plans and in the PR todo checklist :)

nocive commented:
@brianr will add it in a few hours, I still need to briefly check the existing test code and figure out where to do it, but I assume it should be fairly quickly. I'm in Berlin tz tho, so it will have to wait for a new day :)

nocive commented:
@brianr done, added some basic test coverage mimicking ```testMessageWithRequestData()``` method.
Not sure if that's enough or if you guys think something else is needed, let me know.

Also moved ```php_context``` to be part of the base data and to be handled in ```build_base_data()``` instead.

Crisfole commented:
LGTM. Merging later today with a few other ones for a minor version bump.

nocive commented:
@Crisfole awesome, thanks :smiley:

nocive commented:
squashed, rdy for merge

Crisfole commented:
Sorry this took so long!

nocive commented:
@Crisfole better late than never :)

Thanks guys!
